src/compiler/parser.ts

```diff
@@ -1485,6 +1486,11 @@ module ts {
             identifierCount++;
             if (isIdentifier) {
                 let node = <Identifier>createNode(SyntaxKind.Identifier);
+
+                // Store original token kind if it is not just an Identifier so we can report appropriate error later in type checker
+                if (token !== SyntaxKind.Identifier) {
+                    node.originalStrictModeSyntaxKind = token;

```

Comment:
This is probably better named as originalKeywordKind or originalKeywordSyntaxKind.